Resolve hosts for fritzbox_callmonitor

guillempages opened the pull request:
If the configuration supplied "host" is not an IP address, try resolving it

## Description:
I had written a hostname as "host" in my configuration.yaml for the fritzbox_callmonitor platform, and was getting warnings "fritzbox_callmonitor is taking longer than 10 seconds to initialize", and also could not find the phonebook with id 0.

Finally I found out, that I have to pass an IP address, and not a hostname in the host parameter. I personally find it nicer if both can be accepted, so I created this patch that checks whether host is an IP address, and if it isn't, tries to resolve it as a hostname.

**Related issue (if applicable):** I haven't created (or found) an issue for that; I can create one if required. Might be related to #21143.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/11196

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  platform: fritzbox_callmonitor
  username: myFritzboxUser
  password: MY_PASSWORD
  host: fritzbox.example.com
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

guillempages committed:
Resolve hosts for fritzbox_callmonitor

If the configuration supplied "host" is not an IP address, try resolving it

escoand reviewed (commented):
You don't need to check if it's an IP address already:
```
>>> socket.gethostbyname("localhost")
'127.0.0.1'
>>> socket.gethostbyname("127.0.0.1")
'127.0.0.1'
```

guillempages commented:
> You don't need to check if it's an IP address already:
> 
I also thought about that, and checked a little more:
```
>>> socket.gethostbyname("192.168.1.1")
'192.168.1.1'
>>> socket.gethostbyname("192.168")
'192.0.0.168'
```
I would agree, that this is a corner case, but feels wrong to me. If you really want me to change it, I can do it of course; I just see my proposal as more robust.

escoand commented:
> I would agree, that this is a corner case, but feels wrong to me. If you really want me to change it, I can do it of course; I just see my proposal as more robust.

https://www.heise.de/security/meldung/l-f-Warum-1-1-eine-gueltige-IP-Adresse-ist-4470232.html

guillempages committed:
always use gethostbyname

Instead of just checking whether it is an IP and if it isn't try to resolve; just resolve it; IPs will be returned unchanged, and hostnames will be resolved.

guillempages commented:
Ok, you convinced me. I've updated the PR. 

BTW, should I also create a PR for the documentation in home-assistant.io, to specify that it also accepts hostnames?

(P.S. how did you know that I speak German? ;-) )

escoand commented:
> (P.S. how did you know that I speak German? ;-) )

Did just know this German article and found no English on quick googling. ;-)

fabaff commented on homeassistant/components/fritzbox_callmonitor/sensor.py:
```diff
@@ -59,7 +59,8 @@
     name = config.get(CONF_NAME)
-    host = config.get(CONF_HOST)
+    # Try to resolve a hostname; if it is already an IP, it will be returned as-is
+    host = socket.gethostbyname(config.get(CONF_HOST))
```
The documentation says that `host:` should be an IP address. 

Most users don't run their own DNS server at home thus you need to handle the traceback if the lookup fails.

guillempages replied:
If someone uses a hostname, they should at least have configured a "hosts" file, but I see your point. Added exception catching, and using DEFAULT_HOST as fallback. Or would you prefer an error and not initializing the component at all?

Should I also do a PR for the documentation?

guillempages committed:
Catch error if the hostname cannot be resolved

fabaff commented on homeassistant/components/fritzbox_callmonitor/sensor.py:
```diff
@@ -59,8 +59,17 @@
+        host = socket.gethostbyname(host)
+    except socket.error:
+        _LOGGER.warning(
+            "Could not resolve hostname: %s.\nTrying default address %s",
```
Remove the part about the default host. If the lookup fails then the setup should fails. Assuming too much things make it hard to debug for users.

guillempages replied:
done

guillempages committed:
Don't fallback to default host

If the hostname cannot be resolved; don't try to fallback; just print the error message.

fabaff commented on homeassistant/components/fritzbox_callmonitor/sensor.py:
```diff
@@ -64,12 +64,7 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-            DEFAULT_HOST,
-        )
-        host = DEFAULT_HOST
+        _LOGGER.error("Could not resolve hostname %s", host)
```
`setup_platform` needs to fail if the lookup is not successful.

guillempages replied:
Previously it was not failing if the host was not an IP; it just issued a warning, that the phonebook could not be found (and then did not resolve the names), but was being setup correctly, and the call status worked.
I did not want to change the functionality that much, just wanted to allow host names as well as IPs.
Should I change this nevertheless? (might break some installations, that do not care about the phonebook and are (wrongly?) working now).
What's the standard way to fail the setup? return false?

guillempages replied:
I'm now aborting the setup if the hostname cannot be resolved, so that an error is logged and no entities are created.

guillempages committed:
Fail setup if hostname cannot be resolved

If the hostname cannot be resolved, log an error and stop the setup;
no entities will be then created.

cgtobi reviewed (approved):
LGTM
